### ISSUE TITLE
Endless CPU usage after asking to import selected Khan Academy resources

### ISSUE THREAD

**Issue by benjaoming** [learningequality/kolibri#3497]
### Observed behavior

Need more info on this, but this is the best I can produce right now.. observe the 566m of CPU usage:

```
top - 01:39:58 up 2 days, 11:30,  1 user,  load average: 2.10, 2.17, 2.17
Tasks: 287 total,   1 running, 286 sleeping,   0 stopped,   0 zombie
%Cpu(s): 45.6 us, 12.9 sy,  0.0 ni, 41.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :  8054696 total,   904492 free,  5504828 used,  1645376 buff/cache
KiB Swap: 16381948 total, 14482384 free,  1899564 used.  1499508 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                            
15123 kolibri   20   0 3474360 1.112g  10440 S 201.0 14.5 566:35.82 kolibri                         
```

### Steps to reproduce

Not sure, but I have asked to download a select sub-part of the KA resources, and my laptop is stuck for hours.

### Context

Kolibri 0.9.0, debian installation source, Ubuntu 16.04

**Comment by benjaoming** [comment]
Some more debugging: After monitoring bandwidth usage for >10m, the process still hasn't requested anything online. So I would assume that there is some kind of endless loop going on here.

Log doesn't say anything:

```
INFO 2018-04-07 18:41:17,811 cli Going to daemon mode, logging to /var/kolibri/.kolibri/server.log
INFO 2018-04-07 18:41:17,823 ping Attempting a ping.
INFO 2018-04-07 18:41:17,843 ping data: {"language": "en", "sysversion": "3.5.2 (default, Nov 23 2017, 16:37:01) \n[GCC 5.4.0 20160609]", "node_id": "0906dca0fbae53e9abe7", "database_id": "db4903d6f2b84001aaf0ee01ecf45b00", "uptime": 0, "mode": "", "version": "0.9.0", "instance_id": "600bc8a0a69b1dae1d24cf0afa33d34d", "platform": "Linux-4.4.0-119-generic-x86_64-with-Ubuntu-16.04-xenial", "system_id": ""}
```

**Comment by benjaoming** [comment]
```
kolibri@localhost:~$ kolibri start --foreground
INFO     Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO     Running update routines for new version...

0 static files copied to '/var/kolibri/.kolibri/static', 588 unmodified.
Operations to perform:
  Apply all migrations: admin, auth, content, contenttypes, device, exams, kolibriauth, lessons, logger, morango, sessions
Running migrations:
  No migrations to apply.
INFO     Running 'kolibri start' in foreground...
INFO     Kolibri running on:

	http://192.168.1.100:8080/
	http://192.168.1.102:8080/
	http://127.0.0.1:8080/

INFO     Attempting a ping.
INFO     data: {"database_id": "db4903d6f2b84001aaf0ee01ecf45b00", "instance_id": "600bc8a0a69b1dae1d24cf0afa33d34d", "sysversion": "3.5.2 (default, Nov 23 2017, 16:37:01) \n[GCC 5.4.0 20160609]", "platform": "Linux-4.4.0-119-generic-x86_64-with-Ubuntu-16.04-xenial", "mode": "", "system_id": "", "language": "en", "uptime": 0, "version": "0.9.0", "node_id": "0906dca0fbae53e9abe7"}
INFO     Started new MESSAGEPROCESSOR thread ID#thirteen-pluto-don-neptune
INFO     Started new SCHEDULER thread ID#juliet-georgia-eight-social
INFO     Started new WORKERMESSAGEHANDLER thread ID#uranus-zebra-magazine-bulldog
INFO:cherrypy.error:[08/Apr/2018:02:16:40] ENGINE Serving on http://0.0.0.0:8080
INFO:cherrypy.error:[08/Apr/2018:02:16:40] ENGINE Bus STARTED
INFO     Ping succeeded! (response: {}) Sleeping for 1440.0 minutes.
INFO:kolibri.core.analytics.management.commands.ping:Ping succeeded! (response: {}) Sleeping for 1440.0 minutes.
WARNING  Did not find language code None in our le_utils.constants!
WARNING:kolibri.content.api:Did not find language code None in our le_utils.constants!
INFO     Downloading data for channel id 1ceff53605e55bef987d88e0908658c5
INFO:kolibri.content.management.commands.importchannel:Downloading data for channel id 1ceff53605e55bef987d88e0908658c5
100%|███████████████████████████████████████████████████████████████████████████████████████████████| 105951232/105951232 [00:29<00:00, 3535773.07it/s]INFO     Channel 1ceff53605e55bef987d88e0908658c5 already exists in database, cleaning up ContentNodes
INFO:kolibri.content.utils.channel_import:Channel 1ceff53605e55bef987d88e0908658c5 already exists in database, cleaning up ContentNodes
100%|███████████████████████████████████████████████████████████████████████████████████████████████| 105951232/105951232 [00:40<00:00, 3535773.07it/s]INFO     Importing ContentTag data
INFO:kolibri.content.utils.channel_import:Importing ContentTag data
INFO     Importing ContentNode_has_prerequisite data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_has_prerequisite data
INFO     Importing ContentNode_related data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_related data
INFO     Importing ContentNode_tags data
INFO:kolibri.content.utils.channel_import:Importing ContentNode_tags data
INFO     Importing ContentNode data
INFO:kolibri.content.utils.channel_import:Importing ContentNode data
INFO     Importing Language data
INFO:kolibri.content.utils.channel_import:Importing Language data
INFO     Importing File data
INFO:kolibri.content.utils.channel_import:Importing File data
INFO     Importing LocalFile data
INFO:kolibri.content.utils.channel_import:Importing LocalFile data
INFO     Importing AssessmentMetaData data
INFO:kolibri.content.utils.channel_import:Importing AssessmentMetaData data
INFO     Importing ChannelMetadata data
INFO:kolibri.content.utils.channel_import:Importing ChannelMetadata data
INFO     Setting availability of File objects based on LocalFile availability
INFO:kolibri.content.utils.annotation:Setting availability of File objects based on LocalFile availability
INFO     Setting availability of non-topic ContentNode objects based on File availability
INFO:kolibri.content.utils.annotation:Setting availability of non-topic ContentNode objects based on File availability
INFO     Setting availability of ContentNode objects with children for 5 levels
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for 5 levels
INFO     Setting availability of ContentNode objects with children for level 5
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 5
INFO     Setting availability of ContentNode objects with children for level 4
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 4
INFO     Setting availability of ContentNode objects with children for level 3
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 3
INFO     Setting availability of ContentNode objects with children for level 2
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 2
INFO     Setting availability of ContentNode objects with children for level 1
INFO:kolibri.content.utils.annotation:Setting availability of ContentNode objects with children for level 1

```

![image](https://user-images.githubusercontent.com/374612/38461593-d1b41e80-3ad4-11e8-9585-6c06434f01a0.png)

**Comment by benjaoming** [comment]
Can you reproduce this? I sense it happens every time I just select the first 4 top-level items of Khan Academy.. math, science etc.. nothing happens after. In the log, nothing is shown, in the UI, it shows "preparing".

**Comment by jamalex** [comment]
Yeah, I've had this. I think I reported it somewhere too. It's not
*literally* forever for me, but I think it took like an hour or something
once too.

It happens for huge content DBs, with lots of content nodes and/or file
nodes. It's either the tree importation or the availability annotation.
Doesn't always happen. And sometimes it's slow via the web ui and fast via
the command-line import commands. Other times it's also slow via command
line, especially when annotating Level 4 or 5.

tl;dr: more investigation needed -- perhaps the perf team can take a peek
too.

**Comment by bonidjukic** [comment]
> tl;dr: more investigation needed -- perhaps the perf team can take a peek
> too.
ACK, it goes on the list.

**Comment by indirectlylit** [comment]
I believe this has been very slow for large channels ever since we introduced availability annotation

**Comment by kilimba** [comment]
I have the same issue on ubuntu 16.04, UI is stuck on 'prepaing' when trying to import from kolibri studio just the Maths package. Let me know when resolved please.

**Comment by jamalex** [comment]
@kilimba Thanks for the report, and for your patience! We have just released Kolibri v0.9.1, which we expect should resolve this issue (KA importing will still take a minute or two to kick off, due to the massive size of the database entries being loaded, but should no longer take "forever"). You can download it here (and please let us know how it goes): https://learningequality.org/download/

**Comment by benjaoming** [comment]
@kilimba this will be resolved in the next 0.9 patch release, since the fix has been merged to the 0.9 branch.

**Comment by kilimba** [comment]
Guys, splendid job! Upgraded to 0.9.1, works like a charm!! now I can merrily go watch Manchester United thrash Tottenham this afternoon with a load off my mind, thanks again!

**Comment by benjaoming** [comment]
@kilimba hurray! and thanks for the football tip, that might work for me, too! :)

**Comment by rtibbles** [comment]
Glad we could fix the issue - if Man Utd beat Tottenham, I will no longer be glad.

**Comment by jamalex** [comment]
Apparently @kilimba had two wins on Saturday. Sorry, @rtibbles...
